[JIT] Provide more info when attribute fails to convert

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50870 [JIT] Provide more info when attribute fails to convert**

**Summary**
Module attributes whose types cannot be determined based on annotations
or inference based on their values at script time are added to the
concrete type of the corresponding module as "failed attributes". Any
attempt to access them in scripted code produces an error with a message
explaining that the attribute could not be contributed to a
corresponding attribute on the TorchScript module. However, this error
is not more specific than that.

This commit modifies `infer_type` in `_recursive.py` so that it returns
`c10::InferredType` instead, which allows more information about typing
failures to be communicated to the caller through the `reason()` method
on this class. This information is appended to the hint added to the
module concrete type for failed attributes.

**Testing**
This commit adds a unit test to `test_module_containers.py` that checks
that extra information is provided about the reason for the failure
when a module attribute consisting of a list of `torch.nn.Module` fails to convert.

Differential Revision: [D26091472](https://our.internmc.facebook.com/intern/diff/D26091472)

Commit:
[WIP][JIT] Provide more info when attribute fails to convert

**Summary**
Module attributes whose types cannot be determined based on annotations
or inference based on their values at script time are added to the
concrete type of the corresponding module as "failed attributes". Any
attempt to access them in scripted code produces an error with a message
explaining that the attribute could not be contributed to a
corresponding attribute on the TorchScript module. However, this error
is not more specific than that.

This commit modifies `infer_type` in `_recursive.py` so that it returns
`c10::InferredType` instead, which allows more information about typing
failures to be communicated to the caller through the `reason()` method
on this class. This information is appended to the hint added to the
module concrete type for failed attributes.

**Testing**
This commit adds a unit test to `test_module_containers.py` that checks
that extra information is provided about the reason for the failure
when a module attribute consisting of a list of `torch.nn.Module` fails to convert.

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -345,6 +345,17 @@ inline InferredType tryToInferType(py::handle input) {
 #endif
   }
 
+  if (as_module(py::cast<py::object>(input))) {
+    // if obj is already a ScriptModule, just return its ivalue
+    return InferredType("Cannot infer type of ScriptModule");
+  }
```
This is a bit weird. We can obviously infer the type of a `ScriptModule`, but we may not want to do that if it is in a list or a dict, because we don't allow multiple references to the same module. Should this check be moved to `tryToInferContainerType` so that we can still infer the types of individual attributes that happen to be scripted modules correctly?

SplitInfinity commented on torch/jit/_recursive.py:
```diff
@@ -263,15 +268,17 @@ def infer_type(name, item):
-            hint = ("(This attribute exists on the Python module, "
-                    "but we failed to convert Python type: '{}' "
-                    "to a TorchScript type.)").format(torch.typename(type(value)))
+            inferred_msg = "Its type was inferred; try adding a type annotation for the attribute." if inferred else ""
```
I commented about this on the issue but I wasn't able to come up with a test case that can help test this. Under what circumstances does an attribute fail to get converted without an annotation?

SplitInfinity reviewed (commented):
I left some comments about things I want to discuss.

Commit:
Update on "[JIT] Provide more info when attribute fails to convert"


**Summary**
Module attributes whose types cannot be determined based on annotations
or inference based on their values at script time are added to the
concrete type of the corresponding module as "failed attributes". Any
attempt to access them in scripted code produces an error with a message
explaining that the attribute could not be contributed to a
corresponding attribute on the TorchScript module. However, this error
is not more specific than that.

This commit modifies `infer_type` in `_recursive.py` so that it returns
`c10::InferredType` instead, which allows more information about typing
failures to be communicated to the caller through the `reason()` method
on this class. This information is appended to the hint added to the
module concrete type for failed attributes.

**Testing**
This commit adds a unit test to `test_module_containers.py` that checks
that extra information is provided about the reason for the failure
when a module attribute consisting of a list of `torch.nn.Module` fails to convert.

[ghstack-poisoned]

Commit:
Update on "[JIT] Provide more info when attribute fails to convert"


**Summary**
Module attributes whose types cannot be determined based on annotations
or inference based on their values at script time are added to the
concrete type of the corresponding module as "failed attributes". Any
attempt to access them in scripted code produces an error with a message
explaining that the attribute could not be contributed to a
corresponding attribute on the TorchScript module. However, this error
is not more specific than that.

This commit modifies `infer_type` in `_recursive.py` so that it returns
`c10::InferredType` instead, which allows more information about typing
failures to be communicated to the caller through the `reason()` method
on this class. This information is appended to the hint added to the
module concrete type for failed attributes.

**Testing**
This commit adds a unit test to `test_module_containers.py` that checks
that extra information is provided about the reason for the failure
when a module attribute consisting of a list of `torch.nn.Module` fails to convert.

[ghstack-poisoned]

Commit:
Update on "[JIT] Provide more info when attribute fails to convert"


**Summary**
Module attributes whose types cannot be determined based on annotations
or inference based on their values at script time are added to the
concrete type of the corresponding module as "failed attributes". Any
attempt to access them in scripted code produces an error with a message
explaining that the attribute could not be contributed to a
corresponding attribute on the TorchScript module. However, this error
is not more specific than that.

This commit modifies `infer_type` in `_recursive.py` so that it returns
`c10::InferredType` instead, which allows more information about typing
failures to be communicated to the caller through the `reason()` method
on this class. This information is appended to the hint added to the
module concrete type for failed attributes.

**Testing**
This commit adds a unit test to `test_module_containers.py` that checks
that extra information is provided about the reason for the failure
when a module attribute consisting of a list of `torch.nn.Module` fails to convert.

[ghstack-poisoned]

Commit:
Update on "[JIT] Provide more info when attribute fails to convert"


**Summary**
Module attributes whose types cannot be determined based on annotations
or inference based on their values at script time are added to the
concrete type of the corresponding module as "failed attributes". Any
attempt to access them in scripted code produces an error with a message
explaining that the attribute could not be contributed to a
corresponding attribute on the TorchScript module. However, this error
is not more specific than that.

This commit modifies `infer_type` in `_recursive.py` so that it returns
`c10::InferredType` instead, which allows more information about typing
failures to be communicated to the caller through the `reason()` method
on this class. This information is appended to the hint added to the
module concrete type for failed attributes.

**Testing**
This commit adds a unit test to `test_module_containers.py` that checks
that extra information is provided about the reason for the failure
when a module attribute consisting of a list of `torch.nn.Module` fails to convert.

[ghstack-poisoned]

gmagogsfm commented on torch/jit/_recursive.py:
```diff
@@ -118,13 +118,18 @@ def infer_type(name, item):
-        return attr_type
+            inferred = True
+
+        return attr_type, inferred
```
Where is second return value used?

SplitInfinity replied:
See line 277.

gmagogsfm reviewed (approved):
Looks good, just some minor comments

Commit:
Update on "[JIT] Provide more info when attribute fails to convert"


**Summary**
Module attributes whose types cannot be determined based on annotations
or inference based on their values at script time are added to the
concrete type of the corresponding module as "failed attributes". Any
attempt to access them in scripted code produces an error with a message
explaining that the attribute could not be contributed to a
corresponding attribute on the TorchScript module. However, this error
is not more specific than that.

This commit modifies `infer_type` in `_recursive.py` so that it returns
`c10::InferredType` instead, which allows more information about typing
failures to be communicated to the caller through the `reason()` method
on this class. This information is appended to the hint added to the
module concrete type for failed attributes.

**Testing**
This commit adds a unit test to `test_module_containers.py` that checks
that extra information is provided about the reason for the failure
when a module attribute consisting of a list of `torch.nn.Module` fails to convert.

Differential Revision: [D26091472](https://our.internmc.facebook.com/intern/diff/D26091472)

[ghstack-poisoned]

Commit:
Update on "[JIT] Provide more info when attribute fails to convert"


**Summary**
Module attributes whose types cannot be determined based on annotations
or inference based on their values at script time are added to the
concrete type of the corresponding module as "failed attributes". Any
attempt to access them in scripted code produces an error with a message
explaining that the attribute could not be contributed to a
corresponding attribute on the TorchScript module. However, this error
is not more specific than that.

This commit modifies `infer_type` in `_recursive.py` so that it returns
`c10::InferredType` instead, which allows more information about typing
failures to be communicated to the caller through the `reason()` method
on this class. This information is appended to the hint added to the
module concrete type for failed attributes.

**Testing**
This commit adds a unit test to `test_module_containers.py` that checks
that extra information is provided about the reason for the failure
when a module attribute consisting of a list of `torch.nn.Module` fails to convert.

Differential Revision: [D26091472](https://our.internmc.facebook.com/intern/diff/D26091472)

[ghstack-poisoned]
